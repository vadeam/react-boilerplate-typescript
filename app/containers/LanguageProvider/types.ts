import { Record } from 'immutable'

import { LOCALE_EN, LOCALE_DE, DEFAULT_LOCALE } from 'constants/locales'

declare global {
  type Locales = typeof LOCALE_EN | typeof LOCALE_DE
}

export type LocalMessages = { [p: string]: string }

export interface LanguageActionTypes {
  CHANGE_LOCALE: string
  SET_LOCALE_MESSAGES: string
}

export interface ChangeLocaleAction extends Action<LanguageActionTypes> {
  readonly locale: Locales
}

export interface SetLocaleMessagesAction extends Action<LanguageActionTypes> {
  readonly messages: LocalMessages
}

export interface LanguageActionCreators {
  changeLocale: (locale: Locales) => ChangeLocaleAction
  setLocaleMessages: (messages: LocalMessages) => SetLocaleMessagesAction
}

export interface LanguageProviderState extends InheritedReducerState {
  readonly locale: Locales
  readonly messages: LocalMessages
}

export const LanguageProviderStateRecord = Record<LanguageProviderState>(
  {
    locale: DEFAULT_LOCALE,
    messages: {},
  },
  'LanguageProviderStateRecord',
)
