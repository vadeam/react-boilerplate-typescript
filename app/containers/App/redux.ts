import { createActions, createReducer } from 'reduxsauce'
import { Record, get, getIn } from 'immutable'
import { createSelector } from 'reselect'

import { REDUCER_KEY_APP, REDUCER_KEY_ROUTER } from 'constants/reducers'
import { Repo } from 'containers/RepoListItem/types'

import {
  AppActionCreators,
  AppActionTypes,
  AppState,
  AppStateRecord,
  LoadReposAction,
  LoadReposSuccessAction,
  LoadReposErrorAction,
} from './types'

const prefix = `[${REDUCER_KEY_APP}] `
const INITIAL_STATE: Record<AppState> = AppStateRecord()
const { Types, Creators: actions } = createActions<AppActionTypes, AppActionCreators>(
  {
    loadRepos: null,
    reposLoaded: ['repos', 'username'],
    repoLoadingError: ['error'],
  },
  { prefix },
)

const selectAppState = (state: ApplicationRootState) => state[REDUCER_KEY_APP] || INITIAL_STATE
const selectRoute = (state: ApplicationRootState) => state[REDUCER_KEY_ROUTER]
const selectors = {
  selectAppState,
  makeSelectCurrentUser: () => createSelector(selectAppState, (appState) => get(appState, 'currentUser', '')),
  makeSelectFetching: () => createSelector(selectAppState, (appState) => get(appState, 'isFetching', false)),
  makeSelectError: () => createSelector(selectAppState, (appState) => get(appState, 'error', null)),
  makeSelectRepos: () =>
    createSelector(selectAppState, (appState) => getIn(appState, ['userData', 'repos'], null) as Nullable<Repo[]>),
  makeSelectLocation: () => createSelector(selectRoute, (routeState) => routeState.location),
}

const loadRepos: ReducerHandler<Record<AppState>, LoadReposAction> = (state) =>
  state.mergeDeep({
    isFetching: true,
    currentUser: state.get('currentUser'),
    userData: {
      repos: null,
    },
    error: null,
  })
const loadReposSuccess: ReducerHandler<Record<AppState>, LoadReposSuccessAction> = (state, { username, repos }) =>
  state.mergeDeep({
    currentUser: username,
    isFetching: false,
    error: false,
    userData: { repos },
  })
const loadReposError: ReducerHandler<Record<AppState>, LoadReposErrorAction<object>> = (state, { error }) =>
  state.mergeDeep({
    error,
    isFetching: false,
    userData: {
      repos: null,
    },
  })

const reducer = createReducer<Record<AppState>>(INITIAL_STATE, {
  [Types.LOAD_REPOS]: loadRepos,
  [Types.REPOS_LOADED]: loadReposSuccess,
  [Types.REPO_LOADING_ERROR]: loadReposError,
})

export { Types as types, prefix as namespace, INITIAL_STATE, actions, reducer, selectors }
