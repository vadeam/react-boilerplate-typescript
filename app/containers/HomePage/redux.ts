import { createActions, createReducer } from 'reduxsauce'
import { Record, get } from 'immutable'
import { createSelector } from 'reselect'

import { REDUCER_KEY_HOME } from 'constants/reducers'

import { HomeActionCreators, HomeActionTypes, HomeState, HomeStateRecord, HomeAction } from './types'

const prefix = `[${REDUCER_KEY_HOME}] `
const INITIAL_STATE: Record<HomeState> = HomeStateRecord()
const { Types, Creators: actions } = createActions<HomeActionTypes, HomeActionCreators>(
  {
    changeUsername: ['username'],
  },
  { prefix },
)

const selectHome = (state: ApplicationRootState) => state[REDUCER_KEY_HOME] as Record<HomeState>
const selectors = {
  selectHome,
  getUsername: () => createSelector(selectHome, (homeState): string => get(homeState, 'username', '')),
}

const setUsername: ReducerHandler<Record<HomeState>, HomeAction> = (state, { username }) =>
  state.merge({ username: username.replace(/@/gi, '') })

const reducer = createReducer<Record<HomeState>>(INITIAL_STATE, {
  [Types.CHANGE_USERNAME as string]: setUsername,
})

export { Types as types, prefix as namespace, INITIAL_STATE, actions, reducer, selectors }
