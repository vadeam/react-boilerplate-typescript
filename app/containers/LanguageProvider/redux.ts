import { createActions, createReducer } from 'reduxsauce'
import { Record, get } from 'immutable'
import { createSelector } from 'reselect'

import { REDUCER_KEY_LANGUAGE } from 'constants/reducers'
import { DEFAULT_LOCALE } from 'constants/locales'

import {
  LanguageActionCreators,
  LanguageActionTypes,
  LanguageProviderState,
  LanguageProviderStateRecord,
  ChangeLocaleAction,
  SetLocaleMessagesAction,
} from './types'

const prefix = `[${REDUCER_KEY_LANGUAGE}] `
const INITIAL_STATE: Record<LanguageProviderState> = LanguageProviderStateRecord()
const { Types, Creators: actions } = createActions<LanguageActionTypes, LanguageActionCreators>(
  {
    changeLocale: ['locale'],
    setLocaleMessages: ['messages'],
  },
  { prefix },
)

const selectLanguageState = (state: ApplicationRootState) =>
  (state[REDUCER_KEY_LANGUAGE] || INITIAL_STATE) as Record<LanguageProviderState>
const selectors = {
  selectLanguageState,
  makeSelectLanguageProviderState: () =>
    createSelector(selectLanguageState, (languageState) => languageState.toJS() as LanguageProviderState),
  makeSelectLocale: () =>
    createSelector(selectLanguageState, (languageState): Locales => get(languageState, 'locale', DEFAULT_LOCALE)),
}

const changeLocale: ReducerHandler<Record<LanguageProviderState>, ChangeLocaleAction> = (state, { locale }) =>
  state.merge({ locale })
const setLocaleMessages: ReducerHandler<Record<LanguageProviderState>, SetLocaleMessagesAction> = (
  state,
  { messages },
) => state.merge({ messages })

const reducer = createReducer<Record<LanguageProviderState>>(INITIAL_STATE, {
  [Types.CHANGE_LOCALE]: changeLocale,
  [Types.SET_LOCALE_MESSAGES]: setLocaleMessages,
})

export { Types as types, prefix as namespace, INITIAL_STATE, actions, reducer, selectors }
