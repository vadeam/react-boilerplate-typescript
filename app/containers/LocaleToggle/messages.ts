/**
 * LocaleToggle Messages
 * This contains all the text for the LanguageToggle component.
 */
import { defineMessages } from 'react-intl'

import { LOCALE_EN, LOCALE_DE } from 'constants/locales'

export const scope = 'containers.LocaleToggle'

export default defineMessages({
  [LOCALE_EN]: {
    id: `${scope}.en`,
    defaultMessage: 'en',
  },
  [LOCALE_DE]: {
    id: `${scope}.de`,
    defaultMessage: 'de',
  },
})
