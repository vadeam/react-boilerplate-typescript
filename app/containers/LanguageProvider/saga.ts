/**
 * Load/save locale and its messages.
 */

import { takeLatest, call, put } from 'redux-saga/effects'

import { BrowserStorage } from 'lib/classes'
import { CUSTOM_LOCALE_STORAGE_KEY } from 'constants/locales'

import { translationMessages } from '../../i18n'

import { types as LanguageActionTypes, actions } from './redux'

/**
 * Save current locale in browser storage
 */
function* saveLocale(locale) {
  BrowserStorage.getStorage().setItem(CUSTOM_LOCALE_STORAGE_KEY, locale)
}

/**
 * Load messages for locale
 */
function* loadMessages({ locale }) {
  const messages = yield call(translationMessages, locale)

  yield call(saveLocale, locale)
  yield put(actions.setLocaleMessages(messages))
}

/**
 * Root saga manages locale lifecycle
 */
export default function* saga() {
  // @ts-ignore Handles locale data
  yield takeLatest(LanguageActionTypes.CHANGE_LOCALE, loadMessages)
}
