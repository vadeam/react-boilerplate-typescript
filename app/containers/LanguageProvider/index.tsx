/**
 * LanguageProvider
 * this component connects the redux state language locale to the
 * IntlProvider component and i18n messages (loaded from `app/translations`)
 */

import React from 'react'
import { IntlProvider } from 'react-intl'
import { useSelector } from 'react-redux'
import { createSelector } from 'reselect'

import { DEFAULT_LOCALE } from 'constants/locales'
import { REDUCER_KEY_LANGUAGE } from 'constants/reducers'
import { useInjectSaga } from 'utils/redux-injectors'

import { selectors } from './redux'
import saga from './saga'

export interface Props {
  children: React.ReactNode
}

const stateSelector = createSelector(
  selectors.makeSelectLanguageProviderState(),
  (languageProviderState) => languageProviderState,
)

export default function LanguageProvider(props: Props) {
  const { locale, messages } = useSelector(stateSelector)

  useInjectSaga({ saga, key: REDUCER_KEY_LANGUAGE })

  return (
    <IntlProvider locale={locale} messages={messages} defaultLocale={DEFAULT_LOCALE}>
      {React.Children.only(props.children)}
    </IntlProvider>
  )
}
