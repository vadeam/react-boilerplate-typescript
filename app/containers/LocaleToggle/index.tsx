/**
 * LanguageToggle
 */

import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { createSelector } from 'reselect'

import Toggle from 'components/Toggle'
import { appLocales } from 'constants/locales'
import { actions, selectors } from 'containers/LanguageProvider/redux'

import messages from './messages'
import { WrapperStyled } from './styled'

const stateSelector = createSelector(selectors.makeSelectLocale(), (locale) => locale)

export default function LocaleToggle() {
  const locale = useSelector(stateSelector)
  const dispatch = useDispatch()
  const onLocaleToggle = (evt) => dispatch(actions.changeLocale(evt.target.value))

  return (
    <WrapperStyled>
      <Toggle value={locale} values={appLocales} messages={messages} onToggle={onLocaleToggle} />
    </WrapperStyled>
  )
}
