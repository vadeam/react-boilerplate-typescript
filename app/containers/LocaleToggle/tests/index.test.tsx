import { fireEvent, render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'

import { LOCALE_EN, LOCALE_DE } from '../../../constants/locales'
import configureStore from '../../../configureStore'
import history from '../../../utils/history'
import LanguageProvider from '../../LanguageProvider'
import { actions } from '../../LanguageProvider/redux'
import LocaleToggle from '../index'

describe('<LocaleToggle />', () => {
  let store
  const mockedChangeLocale = jest
    .spyOn(actions, 'changeLocale')
    .mockImplementation(() => ({ type: 'test', locale: 'any' } as any))

  beforeAll(() => {
    store = configureStore({}, history)
  })

  it('should match the snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <LanguageProvider>
          <LocaleToggle />
        </LanguageProvider>
      </Provider>,
    )
    expect(container.firstChild).toMatchSnapshot()
    expect(mockedChangeLocale).not.toHaveBeenCalled()
  })

  it(`should present the default '${LOCALE_EN}' english language option`, () => {
    const { queryByDisplayValue } = render(
      <Provider store={store}>
        <LanguageProvider>
          <LocaleToggle />
        </LanguageProvider>
      </Provider>,
    )
    expect(queryByDisplayValue(LOCALE_EN)).toBeInTheDocument()
    expect(mockedChangeLocale).not.toHaveBeenCalled()
  })

  it('should dispatch changeLocale when user selects a new option', () => {
    const { container } = render(
      <Provider store={store}>
        <LanguageProvider>
          <LocaleToggle />
        </LanguageProvider>
      </Provider>,
    )
    const newLocale = LOCALE_DE
    const select = container.querySelector('select')!
    fireEvent.change(select, { target: { value: newLocale } })
    expect(mockedChangeLocale).toHaveBeenCalledWith(newLocale)
  })
})
