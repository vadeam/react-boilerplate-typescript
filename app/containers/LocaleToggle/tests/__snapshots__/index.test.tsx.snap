// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocaleToggle /> should match the snapshot 1`] = `
<div
  class="sc-gsTEea McfQC"
>
  <select
    class="sc-bdfBQB igqybA"
  >
    <option
      value="en"
    >
      en
    </option>
    <option
      value="de"
    >
      de
    </option>
  </select>
</div>
`;
